timemachine/cpp/src/barostat.hpp

```diff
@@ -0,0 +1,88 @@
+// this file implements the MonteCarlo barostat from that of OpenMM
+#pragma once
+
+#include <vector>
+#include <random>
+#include "bound_potential.hpp"
+
+namespace timemachine {
+
+//class Barostat {
+//
+//public:
+//
+//    virtual void inplace_move(
+//        double *d_x,
+//        double *d_box,
+//        const double lambda
+//    );
+//
+//};
+
+class MonteCarloBarostat{
+
+
+public:
+
+    MonteCarloBarostat(
+        const int N,
+        const double pressure, // in bar
+        const double temperature, // in kelvin
+        std::vector<std::vector<int> > group_idxs,
+        const int interval,
+        std::vector<BoundPotential *> bps,
+        const int seed
+    );
+
+
+    ~MonteCarloBarostat();
+
+    // inplace_move() may modify d_x and d_box
+    void inplace_move(
+        double *d_x,
+        double *d_box,
+        const double lambda
+    );
+
+    void set_interval(const int interval);
+
+    void set_pressure(const double pressure);
+
+
+private:
+
+    void reset_counters();
+
+    const std::vector<BoundPotential *> bps_;
+
+    const int N_;
+    double pressure_;
+    const double temperature_;
+    int interval_;
+    const int seed_;
+    const std::vector<std::vector<int> > group_idxs_;
+
+    // stuff that deals with RNG
+    std::mt19937 mt_; // Expected to be seeded with seed_
+    std::uniform_real_distribution<double> dist_; // guaranteed unbiased
```

Comment:
Question: should this use the same RNG approach as the thermostat? https://github.com/proteneer/timemachine/blob/6658dde849574fe4b5b6eb3d84e7b9422bd20deb/timemachine/cpp/src/integrator.cu#L29-L31

The barostat consumes far less randomness than the thermostat so I think this is fine either way.

Reply:
Since as it stands the random variables are used in C++ space, the C++ RNG seemed better suited. It has about 2 microsecond over head for the two random numbers generated per move, which is only a small fraction of the total time.